src/TRoom.cpp

```diff
@@ -312,6 +338,86 @@ bool TRoom::hasExit(int direction)
     return false;
 }
 
+// Confirms whether the text identifies a valid exit - this is a bit of a dog's
+// dinner because different parts of this class represent exit directions in
+// different ways...!
+bool TRoom::hasExitOrSpecialExit(const QString & text, const bool isForCustomExitLines = false) const
+{
+    // First check the special exits:
+    QMapIterator<int, QString> itSpecialExit(other);
+    while (itSpecialExit.hasNext()) {
+        itSpecialExit.next();
+        QString exitCmd = itSpecialExit.value();
+        if (exitCmd.startsWith(QLatin1Char('0')) || exitCmd.startsWith(QLatin1Char('1'))) {
+            exitCmd.remove(0, 1);
+        }
+
+        if (exitCmd == text) {
+            // We have a special exit which matches the given text
+            return true;
+        }
+    }
+
+    // Now check the normal ones:
+    if (isForCustomExitLines) {
+        if (text == QLatin1String("N")) {
```

Comment:
You say *most games aren't* but we have to cover *all* games and because (up to the present and will be unless it gets changed**¹**) the text for a special exit represents the *exact* text that has to be sent to the MUD Server to *use* that Special exit we have to reflect that sensitivity in distinguishing exit identifiers.

Parts of the mapper (or rather the Rooms therein) ARE case sensitive - which is down to the (borked IMO) decision to **just** use strings as keys to the `QMap`/`QHash`s that hold details about custom exit lines, door markers and exit weights. The first of these uses CAPITALS as normal exit directions, the others used lower case. Of course - all of these ALSO support special exits as well (which is why text is used) - whereas other map features such as exit locks, and stub exits use numeric codes but can thus only support normal exits and does not work for special exits! That is why special exit locks are provided by the unholy mess that puts a '0' or '1' as a ***prefix*** onto the special exit text when it is stored as the `value` in the `(QMap<int, QString>) TRoom::other` member - which makes searching/iterating through special exits such a PITA.

This mess is how it is right now.

I want (and have played with code) to change to use `QPair<quint8, QString>` as a uniform exit identifier - with the `first` being the existing exit codes (1-12, 13 for Special) and `second` being `QString()` for the normal exit directions or the text for the special case. In some situations it ought to be allowable for some parts of the relevant code to allow `first` to carry the *translated* text for the normal exit direction but that would have to be worked out carefully...

*1 - Long-term I have a plan to allow a short identifier, i.e. a* name *to be assigned to a Special exit which will be used for **display** purposes and then move the current text to be the* command *- with the default when importing old maps to set them both to be the current text - so that, for example, a current `unlock Grate; open Grate; climb down; lock Grate` could be given a name `grate` which would be much easier to use to identify that exit. Whilst this partially looks like an* alias *it would be carried in the Map data, specifically in the Room details. FTR The methodology is similar to how* TinTin++ *does it - although that client does this name/command separation for* all *exits - it is just that the* normal *directions have a default name-command mapping of 'n' to 'n', 'ne' to 'ne' etc.*

Reply:
Okay, my point here is - should not be forced to give normal cardinal directions in all caps. Those should be case-insensitive. I don't agree on sensitive special exits either, but we can let it be.

Reply:
I think I see what you are saying but we ideally we can't be case-sensitive for just the Special exit directions on (just those *two* commands) because we won't be able to tell in current map save files whether, for a custom exit line an `"e"` is a Special exit or a Normal one - OTOH this is broken corner case territory anyhow as for exit weights the same applies (in the opposite direction) for `"E"`... :man_shrugging:

Given that it is already borked for *this* PR I'll propose making map format 20 and later:
+ handle both of these cases in a case insensitive manner.
+ convert the custom exit line keys (and the internal code) to be lower case for the keys in the custom line data held in the `TRoom` class (and remove the case-sensitivity for both `setExitWeight` and `addCustomLine` exit direction *as a string*).
+ revise the Wiki to note the change and (if it is not already there but I think it is) the existing ambiguity between Normal and Special exit directions.

Then - for a separate PR - revise the `TRoom` data structures to be more agnostic about whether the exit is Special or not.  This would have the benefits, if done right, of:
+ allowing Special stub exits - i.e. for noting where there is such an exit but where the destination is not known during mapping.
+ probably allowing simpler C++ code when it is necessary to examine *all* of a room's exits.
+ it will also allow the exit lock data for Special exits to be handled in the same manner and simplify code in that area - and finally allow me to fixup or even eliminate `TRoom::other`.

After a review of existing lua commands that handle exit directions it may be desirable to declare some of them *deprecated* (but still supported, obviously) and provide replacements that have a uniform means of specifying the exit direction {in the manner I have previously mentioned, as a `number`(1-12) + `nil` (or rather ignored) for normal exits and `13` + `string` for special ones}...